README.md
Dredd::Rack
===========

[![Gem Version](https://badge.fury.io/rb/dredd-rack.svg)](http://badge.fury.io/rb/dredd-rack)
[![Build Status](https://travis-ci.org/gonzalo-bulnes/dredd-rack.svg?branch=master)](https://travis-ci.org/gonzalo-bulnes/dredd-rack)
[![Code Climate](https://codeclimate.com/github/gonzalo-bulnes/dredd-rack.svg)](https://codeclimate.com/github/gonzalo-bulnes/dredd-rack)
[![Dredd Reference Version](https://img.shields.io/badge/dredd_reference_version-2.2.5-brightgreen.svg)](https://github.com/apiaryio/dredd)
[![Inline docs](http://inch-ci.org/github/gonzalo-bulnes/dredd-rack.svg?branch=master)](http://inch-ci.org/github/gonzalo-bulnes/dredd-rack)

This gem provides a [Dredd][dredd] runner and a `dredd` rake task to make [API blueprint][blueprint] testing convenient for Rack applications. When verifying blueprints locally, an application server is automatically set up, without need of any manual intervention.

Besides being convenient, that allows to use the API blueprints as acceptance test suites to practice [BDD][rspec-book] with Dredd and RSpec, for example, while clients developers use [Apiary][apiary] as a mock server.

  [dredd]: https://github.com/apiaryio/dredd
  [blueprint]: https://apiblueprint.org/
  [rspec-book]: https://pragprog.com/book/achbd/the-rspec-book
  [apiary]: http://apiary.io
  [issues]: https://github.com/gonzalo-bulnes/dredd-rack/issues
  [gonzalo-bulnes]: https://github.com/gonzalo-bulnes

Installation
------------

Add the gem to your `Gemfile`:

```ruby
# Gemfile

gem 'dredd-rack', '~> 1.0' # see semver.org
```

Define which application Dredd::Rack must serve automatically:

```ruby
# config/initializers/dredd-rack.rb or app.rb or Rakefile

require 'dredd/rack'

Dredd::Rack.configure do |config|
  # Allow the automatic setup of a local application server when necessary
  #
  # Find the name of your application in its `config.ru` file.
  config.app = Example::Application # or Rails.application, Sinatra::Application...

  # Optionally, you can define a custom Dredd command:
  config.dredd_command = 'dredd'
end

```

Usage
-----

### Getting started

Define the `dredd` rake task in your `Rakefile`:

```ruby
# Rakefile

require 'dredd/rack'

# ...

# Define the :dredd task
Dredd::Rack::RakeTask.new # run: `dredd doc/*.apib doc/*.apib.md <local or remote URL>`

# Optionally add the API blueprint verification to the default test suite
# task :default => [:spec, :dredd]
```

Run the API blueprint verification:

```bash
# locally
rake dredd

# or specify a remote server:
#API_HOST=http://api.example.com rake dredd
```

The `:dredd` rake task requires no further configuration. However, it relies on the following convention: API blueprints must be stored in the `doc` directory with either the `.apib` or the `.apib.md` extension.

### Further usage

Of course you can define your own rake tasks with custom runners!

Each Dredd::Rack runner stores a specific Dredd configuration. In order to manage different runners, you can define multiple rake tasks with custom names or descriptions:

```ruby
# Rakefile

# ...

namespace :blueprint do
  namespace :verify do

    desc 'Verify the whole API compliance with its blueprint'
    Dredd::Rack::RakeTask.new(:all)

    desc 'Verify the Machines API compliance with its blueprint'
    Dredd::Rack::RakeTask.new(:machines) do |task|
      task.runner.configure do |dredd|
        dredd.only 'Machines > Machines Collection > List all Machines',
                   'Machines > Machine > Retrieve a Machine' # etc.
        dredd.sorted!.details!
      end
    end

    # ...
  end
end

# run with `rake blueprint:verify:all` and `rake blueprint:verify:machines`
```

Both the [rake task documentation][rake-task-doc] and the [runner documentation][runner-doc] contain more examples, and the runner documentation does also contain a comprehensive list of the [configuration methods][conf] that can be used to customize runners. Take a look at them!

  [rake-task-doc]: doc/rake_task.md
  [runner-doc]: doc/runner.md
  [conf]: doc/runner.md#configuration-methods-reference


#### Remote API testing

Configuring a runner for remote API blueprint compliance testing is as easy as setting its **api_endpoint** option (see the [runner documentation][conf]). Please remember that you must ensure that the API is being served at the specified URI when performing remote API blueprint validation.

Development
-----------

### Testing and documentation


This gem behaviour is described using [RSpec][rspec] and RSpec [tags][tags] are used to categorize the spec examples.

Spec examples that are tagged as `public` describe aspects of the gem public API, and MAY be considered as the gem documentation.

The `private` or `protected` specs are written for development purpose only. Because they describe internal behaviour which may change at any moment without notice, they are only executed as a secondary task by the [continuous integration service][travis] and SHOULD be ignored.

Run `rake spec:public` to print the gem public documentation.

  [rspec]: https://www.relishapp.com/rspec/rspec-rails/docs
  [tags]: https://www.relishapp.com/rspec/rspec-core/v/3-1/docs/command-line/tag-option
  [travis]: https://travis-ci.org/gonzalo-bulnes/dredd-rack/builds

Contributions
-------------

Contributions are welcome! The best way to get in touch is probably to open an issue, whether it is to get help, give feedback or discuss an idea. Also, when there are plan to extend Dredd::Rack, the [intended API][intended] is [described beforehand so we can discuss it][rdd]. Feel free to jump into the discussion at any moment!

Note that opening pull requests with work in progress is not only welcome, but encouraged! Let's talk early, talk often, and we will all learn in the way :)

Please note that Dredd::Rack is released with a [Contributor Code of Conduct][coc]. By participating in this project you agree to abide by its terms.

  [rdd]: http://tom.preston-werner.com/2010/08/23/readme-driven-development.html
  [intended]: ./doc/README.md
  [coc]: ./CODE_OF_CONDUCT.md

Credits
-------

The `Dredd::Rack::RakeTask` is heavily inspired in the [`RSpec::Core::RakeTask`][rspec-core-raketask]. Let the corresponding credits go to the RSpec team.

  [rspec-core-raketask]: https://github.com/rspec/rspec-core/blob/v3.2.1/lib/rspec/core/rake_task.rb

License
-------

    Dredd::Rack provides convenient API blueprint testing to Rack applications.
    Copyright (C) 2015, 2016  Gonzalo Bulnes Guilpain

    Copyright (C) 2012 Chad Humphries, David Chelimsky, Myron Marston
    Copyright (C) 2009 Chad Humphries, David Chelimsky
    Copyright (C) 2006 David Chelimsky, The RSpec Development Team
    Copyright (C) 2005 Steven Baker

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2019-06-12

### Added

- Support for Capybara 2 and 3 - @jesperronn

### Fixed

- Remove `Gemfile.lock` - mostly to acknowledge that it is only used in development and is not really necessary.

## [0.11.0] - 2016-12-29

### Added

- A Contributor Code of Conduct to ensure everyone feels safe contributing
- Add support for Rake 12 - @tobiashm
- Add support for Dredd v2.2.5

## [0.10.0] - 2016-08-06

### Added

- Add support for Rake 11 - @tobiashm
- Add support for Dredd v1.5.0

## [0.9.0] - 2016-06-15

### Added

- Add **dredd_command** option - with help from @gottfrois
- Add configuration handler for convenient initializers

## [0.8.2] - 2016-06-08

### Fixed

- Fix inaccurate counting of command arguments for local API - @rylnd
- Minor fix missing documentation for Dredd v1.0.8 options

## [0.8.1] - 2016-05-07

### Added

- Add support for Dredd v1.0.8
- This change log : )

## [0.8.0] - 2016-04-11 [YANKED]

## [0.7.1] - 2015-12-15

### Fixed

- Fix missing quotes around Dredd CLI arguments

## [0.7.0] - 2015-08-06

### Added

- Provide feedback on invalid command errors - @Maxim-Filimonov

## [0.6.0] - 2015-08-06

### Added

- Add support for Dredd v1.0.0

## [0.5.0] - 2015-05-13

### Added

- Add support for Rails initializers - with help from @rylnd

## [0.4.0] - 2015-03-23

### Added

- Add runner configuration support to rake tasks

## [0.3.0] - 2015-03-16

### Added

- Add clonable Rake task as RSpec does

## [0.2.0] - 2015-02-27

### Added

- Starting a server manually is no longer required

## 0.1.0 - 2015-02-27

### Added

- Allow to run Dredd from a Rake task

## Previously

The original implementation of the Rake task was shared in this [gist][gist].

[1.0.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.11.0...v1.0.0
[0.11.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.10.0...v0.11.0
[0.10.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.9.0...v0.10.0
[0.9.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.8.2...v0.9.0
[0.8.2]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.8.1...v0.8.2
[0.8.1]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.7.1...v0.8.1
[0.8.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.7.1...v0.8.0
[0.7.1]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.7.0...v0.7.1
[0.7.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/gonzalo-bulnes/dredd-rack/compare/v0.1.0...v0.2.0
[gist]: https://gist.github.com/gonzalo-bulnes/eec3f73cc7d6605add21

## Inspiration

Thanks to @nTraum for pointing me at http://keepachangelog.com and to @olivierlacan for writing it in the first place!


spec/spec_helper.rb
require 'bundler/setup'

Bundler.setup

require 'dredd/rack'

Dir["./spec/support/**/*.rb"].sort.each { |f| require f; puts f }

Capybara.server = :webrick

# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # make sure the deprecated RSpec 2 syntax is not used
  config.raise_errors_for_deprecations!

  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end
end


spec/lib/dredd/rack_spec.rb
require 'spec_helper'

describe Dredd::Rack do

  it 'is configurable', public: true do
    expect(subject).to be_kind_of Dredd::Rack::Configuration
  end
end


spec/lib/dredd/runner_spec.rb
require 'spec_helper'

describe Dredd::Rack::Runner do

  describe '#run' do

    describe 'with distinct arguments sets' do

      context 'with the sets of arguments from sample 1' do
        it 'runs Dredd' do
          allow(subject).to receive(:system)
          expect(subject).to receive(:system).with('dredd <path to blueprint> <another path to blueprint> <api_endpoint>')
          subject.run(['<path to blueprint>', '<another path to blueprint>'], '<api_endpoint>')
        end
      end

      context 'with the sets of arguments from sample 2' do
        it 'runs Dredd' do
          allow(subject).to receive(:system)
          expect(subject).to receive(:system).with('dredd <api_endpoint> --dry-run --sorted')
          subject.run([], '<api_endpoint>', dry_run: true, sorted: true)
        end
      end

      context 'with the sets of arguments from sample 3' do
        it 'runs Dredd' do
          allow(subject).to receive(:system)
          expect(subject).to receive(:system).with('dredd <api_endpoint> --dry-run --level info')
          subject.run([], '<api_endpoint>', dry_run: true, level: :info)
        end
      end

      context 'with the sets of arguments from sample 4' do
        it 'runs Dredd' do
          allow(subject).to receive(:system)
          expect(subject).to receive(:system).with('dredd <path to blueprint> <api_endpoint>')
          subject.run(['<path to blueprint>'], '<api_endpoint>')
        end
      end

      context 'with the sets of arguments from sample 5' do
        it 'runs Dredd' do
          allow(subject).to receive(:system)
          expect(subject).to receive(:system).with("dredd <path to blueprint> <api_endpoint> --only 'Machines > Machines collection > Get Machines' --only 'Machine > Get Machines'")
          subject.run(['<path to blueprint>'], '<api_endpoint>', only: ['Machines > Machines collection > Get Machines', 'Machine > Get Machines'])
        end
      end
    end
  end
end


lib/dredd/rack/rake_task.rb
require 'capybara'
require 'rainbow'
require 'rake'
require 'rake/tasklib'

# This class is heavily inspired in the RSpec::Core::RakeTask, which
# was published under the MIT license by:
#
# Copyright (C) 2009 Chad Humphries, David Chelimsky
# Copyright (C) 2006 David Chelimsky, The RSpec Development Team
# Copyright (C) 2005 Steven Baker
#
# See https://github.com/rspec/rspec-core/blob/v3.2.2/lib/rspec/core/rake_task.rb
#
# Modifications are part of Dredd::Rack, published under the GNU GPLv3 ot later by:
#
# Copyright (C) 2015 Gonzalo Bulnes Guilpain
#
# See https://github.com/gonzalo-bulnes/dredd-rack/blob/v0.3.0/lib/dredd/rack/rake_task.rb

module Dredd
  module Rack

    # A clonable Rake task powered by a Dredd::Rack::Runner
    #
    # Examples:
    #
    #    require 'dredd/rack'
    #    Dredd::Rack::RakeTask.new # run it with `rake dredd`
    #
    #    # Customize the name or description of the Rake task:
    #    namespace :blueprint do
    #      desc 'Verify an API complies with its blueprint'
    #      Dredd::Rack::RakeTask.new(:verify)
    #    end
    #    # run it with `rake blueprint:verify`
    #
    class RakeTask < ::Rake::TaskLib

      # Return the task's name
      attr_reader :name

      # Return the task's description
      attr_reader :description

      # Return the Dredd::Rack::Runner instance
      attr_reader :runner

      # Define a task with a custom name, arguments and description
      def initialize(*args, &task_block)
        @name = args.shift || :dredd
        @description = 'Run Dredd::Rack API blueprint verification'
        @runner = Dredd::Rack::Runner.new(ENV['API_HOST'])

        desc description unless ::Rake.application.last_description
        rake_task = task name, *args do |task_args|
          task_block.call(*[self, task_args].slice(0, task_block.arity)) if task_block
          run_task(runner)
        end

        integrate_with_rails!(rake_task)
      end

      private

        def dredd_available?
          system ['which', Dredd::Rack.dredd_command].join(' ')
        end

        def dredd_connection_error?(exit_status)
          exit_status == 8
        end

        def command_message(runner)
          <<-eos.gsub /^( |\t)+/, ""
            #{runner.command}

          eos
        end

        def connection_error_message(runner)
          <<-eos.gsub /^( |\t)+/, ""

            #{Rainbow("Something went wrong.").red}
            Maybe your API is not being served at #{runner.api_endpoint}?

            Note that specifying a different host is easy:
            #{Rainbow('`rake blueprint:verify API_HOST=http://localhost:4567`').yellow}

          eos
        end

        def dredd_not_available_message
          <<-eos.gsub /^( |\t)+/, ""

            The #{Rainbow('dredd').red} blueprint testing tool is not available.
            You may want to install it in order to validate the API blueprints.

            Try #{Rainbow('`npm install dredd --global`').yellow} (use `sudo` if necessary)
            or see https://github.com/apiaryio/dredd for instructions.

          eos
        end

        def run_task(runner)
          abort dredd_not_available_message unless dredd_available?

          puts starting_message

          puts command_message(runner)

          success = runner.run
          exit_status = $?.exitstatus

          puts connection_error_message(runner) unless success if dredd_connection_error?(exit_status)

          abort unless exit_status == 0
        end

        def starting_message
          <<-eos.gsub /^( |\t)+/, ""

            #{Rainbow('Verify the API conformance against its blueprint.').blue}
          eos
        end

        def integrate_with_rails!(rake_task)
          rake_task.enhance([:environment]) if Rake::Task.task_defined? :environment
        end
    end
  end
end


lib/dredd/rack/configuration.rb
module Dredd
  module Rack

    # Hold the Dredd::Rack global configuration.
    module Configuration

      # Return the application to be tested locally
      def app
        @@app
      end

      # Set the application to be tested locally
      #
      # Any Dredd::Rack::Runner configured to run locally will serve
      # this application from a Capybara::Server instance.
      #
      # object - the application Constant
      def app=(object)
        @@app = object
      end

      # Return the command to be runned to invoke Dredd
      def dredd_command
        @@dredd_command
      end

      # Set a custom Dredd command
      #
      # command - the command String
      def dredd_command=(command)
        @@dredd_command = command
      end

      # Default configuration
      @@app = nil
      @@dredd_command = 'dredd'

      # Allow the default configuration to be overwritten from initializers
      def configure
        yield self if block_given?
      end
    end
  end
end


lib/dredd/rack/runner.rb
require 'capybara'

module Dredd
  module Rack

    # A Ruby wrapper around the Dredd API blueprint validation tool
    #
    # Usage:
    #
    #    # run `dredd doc/*.apib doc/*.apib.md http://localhost:XXXX --level warning --dry-run`
    #    # You don't need to start any server, Dredd::Rack does it for you.
    #    dredd = Dredd::Rack::Runner.new
    #    dredd.level(:warning).dry_run!.run
    #
    #    # You can specify an API endpoint to perform a remote validation.
    #    # Do not forget to serve the API at the given URL!
    #    #
    #    # runs `dredd blueprints/*.md doc/*.md https://api.example.com --no-color`
    #    anderson = Anderson::Rack::Runner.new 'https://api.example.com' do |options|
    #      options.paths_to_blueprints 'blueprints/*.md', 'doc/*.md'
    #      options.no_color!
    #    end
    #    anderson.run
    #
    class Runner

      undef_method :method

      NEGATABLE_BOOLEAN_OPTIONS = [:dry_run!, :sandbox!, :names!, :init!, :sorted!, :inline_errors!,
                                   :details!, :color!, :timestamp!, :silent!]
      META_OPTIONS              = [:help, :version]
      BOOLEAN_OPTIONS           = NEGATABLE_BOOLEAN_OPTIONS + META_OPTIONS

      SINGLE_ARGUMENT_OPTIONS   = [:hookfiles, :language, :server, :server_wait, :custom, :only,
                                   :reporter, :output, :header, :user, :method, :level, :path,
                                   :hooks_worker_timeout, :hooks_worker_connect_timeout,
                                   :hooks_worker_connect_retry, :hooks_worker_after_connect_wait,
                                   :hooks_worker_term_timeout, :hooks_worker_term_retry,
                                   :hooks_worker_handler_host, :hooks_worker_handler_port,
                                   :config]
      OPTIONS                   = BOOLEAN_OPTIONS + SINGLE_ARGUMENT_OPTIONS

      # Store the Dredd command line options
      attr_accessor :command_parts

      # Initialize a runner instance
      #
      # The API endpoint can be local or remote.
      #
      # api_endpoint - the API URL as a String
      #
      def initialize(api_endpoint=nil)
        @dredd_command = Dredd::Rack.dredd_command
        @paths_to_blueprints = 'doc/*.apib doc/*.apib.md'
        @api_endpoint = api_endpoint || ''
        @command_parts = []

        yield self if block_given?
      end

      # Set or return the runner API endpoint
      #
      # Use with no arguments to read the runner API endpoint,
      # provide an API endpoint to set it.
      #
      # api_endpoint - String URL of the API endpoint to validate
      #
      # Returns the String URL of the runner API endpoint.
      def api_endpoint(api_endpoint=nil)
        @api_endpoint = api_endpoint unless api_endpoint.nil?
        @api_endpoint
      end

      # Return the Dredd command line
      def command
        ([@dredd_command, @paths_to_blueprints, @api_endpoint] + @command_parts).join(' ')
      end

      # Configure the runner instance
      def configure
        yield self if block_given?
      end

      # Define custom paths to blueprints
      #
      # paths_to_blueprints - as many Strings as paths where blueprints are located
      #
      # Returns self.
      def paths_to_blueprints(*paths_to_blueprints)
        raise ArgumentError, 'invalid path to blueprints' if paths_to_blueprints == [''] || paths_to_blueprints.empty?

        @paths_to_blueprints = paths_to_blueprints.join(' ')
        self
      end

      # Run Dredd
      #
      # Returns true if the Dredd exit status is zero, false instead.
      def run
        raise InvalidCommandError.new(command) unless command_valid?
        start_server! unless api_remote?
        Kernel.system(command)
      end

      # Ensure that the runner does respond_to? its option methods
      #
      # See http://ruby-doc.org/core-2.2.0/Object.html#method-i-respond_to_missing-3F
      def respond_to_missing?(method, include_private=false)
        OPTIONS.include?(method.to_sym ) ||
        NEGATABLE_BOOLEAN_OPTIONS.include?(method.to_s.gsub(/\Ano_/, '').to_sym) ||
        super
      end

      private

        def api_remote?
          !@api_endpoint.empty?
        end

        def command_valid?
          if api_remote?
            command.has_at_least_two_arguments?
          else
            command.has_at_least_one_argument?
          end
        end

        def start_server!
          server = Capybara::Server.new(Dredd::Rack.app)
          server.boot
          @api_endpoint = "http://#{server.host}:#{server.port.to_s}"
        end

        # Private: Define as many setter methods as there are Dredd options
        #
        # The behaviour of Object#method_missing is not modified unless
        # the called method name matches one of the Dredd options.
        #
        # name - Symbol for the method called
        # args - arguments of the called method
        #
        # See also: http://ruby-doc.org/core-2.2.0/BasicObject.html#method-i-method_missing
        def method_missing(name, *args)
          super unless OPTIONS.include?(name.to_sym ) ||
                       NEGATABLE_BOOLEAN_OPTIONS.include?(name.to_s.gsub(/\Ano_/, '').to_sym)

          option_flag = name.to_s.gsub('_', '-').gsub('!', '').prepend('--')
          command_parts = self.command_parts.push option_flag
          command_parts = self.command_parts.push args.slice(0).to_s.quote! if SINGLE_ARGUMENT_OPTIONS.include? name
          self
        end

    end
  end
end

class String

  # Verify that a command has at least one argument (excluding options)
  #
  # Examples:
  #
  #    "dredd doc/*.apib".valid? # => true
  #    "dredd doc/*.apib doc/*apib.md".valid? # => true
  #    "dredd doc/*.apib --level verbose".valid? # => true
  #    "dredd".valid? # => false
  #    "dredd --dry-run".valid? # => false
  #    "dredd --dry-run --level verbose".valid? # => false
  #
  # Known limitations:
  #
  # Does not support short flags. (e.g. using `-l` instead of `--level`).
  # Requires options to be specified after the last argument.
  #
  # Note:
  #
  # The known limitations imply that there may be false negatives: this method
  # can return false for commands that do have two arguments or more. But there
  # should not be false positives: if the method returns true, then the command
  # does have at least two arguments.
  #
  # Returns true if the command String has at least one arguments, false otherwise.
  def has_at_least_one_argument?
    split('--').first.split(' ').length >= 2
  end

  # Verify that a command has at least two arguments (excluding options)
  #
  # Examples:
  #
  #    "dredd doc/*.apib http://api.example.com".valid? # => true
  #    "dredd doc/*.apib doc/*apib.md http://api.example.com".valid? # => true
  #    "dredd doc/*.apib http://api.example.com --level verbose".valid? # => true
  #    "dredd http://api.example.com".valid? # => false
  #    "dredd doc/*.apib --dry-run".valid? # => false
  #    "dredd --dry-run --level verbose".valid? # => false
  #
  # Known limitations:
  #
  # Does not support short flags. (e.g. using `-l` instead of `--level`).
  # Requires options to be specified after the last argument.
  #
  # Note:
  #
  # The known limitations imply that there may be false negatives: this method
  # can return false for commands that do have two arguments or more. But there
  # should not be false positives: if the method returns true, then the command
  # does have at least two arguments.
  #
  # Returns true if the command String has at least two arguments, false otherwise.
  def has_at_least_two_arguments?
    split('--').first.split(' ').length >= 3
  end

  # Include quotes as part of the string
  #
  # Examples:
  #
  #    "Hello, world!".quote! # => "\"Hello, world!\""
  #    "Hello, world!".size # => 13
  #    "Hello, world!".quote!.size # => 15
  #
  # Returns a String, whose first and last characters are quotes.
  def quote!
    '"' + self + '"'
  end
end

Anderson = Dredd # Anderson::Rack::Runner.new runs just as fast as Dredd
